### ISSUE TITLE
Need for proper mention in the documentation  that accelerate.cpu_offload   can not be used for training

### ISSUE THREAD

**Issue by CO18326** [huggingface/accelerate#3770]
During our research while  experimenting with accelerate.cpu_offload  we figured out that it can not be used for training although the training script runs without any error.

1. The mechanism the accelerate.cpu_offload  uses is the reason for it, as it maintains the weight_map dict in the cpu side which stores the original parameter of model
2. During forward pass it loads those values to execution device.
3. Hence computation graph hold the refrences to the current tensors that are on GPU
4. After forward pass model parameters are moved to meta.
5. optimiser.step tries to updates the model.parameters but it holds the references to old model parameters objects when model was initlised, and not see any gradient writen on them, hence no update.

-> conclusion : forward pass is happening using different parameters, gradient is written on different parameters, and optimiser.step is happening on different parameter.
->Interesting point is script dont give any error.

  `  
   
    model = AutoModelForCausalLM.from_pretrained(model_name)
    optimizer = AdamW(model.parameters(), lr=5e-5)
    model = cpu_offload(model, torch.device("cuda"))

    dataset = load_dataset("wikitext", "wikitext-2-v1", split="train")

    def tokenize_fn(examples):
        model_inputs = tokenizer(
            examples["text"],
            truncation=True,
            padding="max_length",
            max_length=args.seq_len,
        )
        labels = []
        for seq in model_inputs["input_ids"]:
            labels.append([tok if tok != tokenizer.pad_token_id else -100 for tok in seq])
        model_inputs["labels"] = labels
        return model_inputs

    dataset = dataset.map(tokenize_fn, batched=True, remove_columns=["text"])
    dataloader = DataLoader(dataset, batch_size=2, shuffle=True, collate_fn=default_data_collator)
    optimizer, dataloader = accelerator.prepare(optimizer, dataloader)

    model.train()
    for step, batch in enumerate(dataloader):
        outputs = model(**batch)
        loss = outputs.loss
        loss.backward()
        optimizer.step()
        optimizer.zero_grad()
        if step + 1 >= args.steps:
            break`

By logging the parameter objects referenced by optimiser , it can be validated that gradients are none , and parameter values are not updating:
<img width="1221" height="366" alt="Image" src="https://github.com/user-attachments/assets/1e0cff9e-caa5-46f2-90a5-efb6501e732c" />

The following script demonstrates what is essentially happening 

[test_script](https://github.com/user-attachments/files/22250322/test-test.py)

`

**Comment by S1ro1** [comment]
cc @SunMarc

**Comment by SunMarc** [comment]
Indeed, `cpu_offload` shouldn't be used for training ! Would you mind updating the docs for this @CO18326 ?

**Comment by CO18326** [comment]
sure

**Comment by CO18326** [comment]
sure i will, can this be treated as bug as well, or it is intended to be used for inference only from the begning ... @SunMarc

**Comment by SunMarc** [comment]
It is only intended to be used for inference from the beginning !
